### PR TITLE
Add support for nested struct pointers #87

### PR DESCRIPTION
**Pull request by frozzare**
This pull request will solve the panic that happens with nested struct pointers as reported in #87. It's solved by using the same idea as for nested structs but for nested struct pointers, also adding the test cases that was provided in the issue and it looks the same as the test case for nested structs. Using a bit of reflection magic to solve the issue :)

### CODE REVIEW

File: encoder_test.go [hunk 1]
```diff
@@ -330,3 +330,16 @@ func TestEncoderWithOmitempty(t *testing.T) {
 	valNotExists(t, "f08", vals)
 	valsExist(t, "f09", []string{"test"}, vals)
 }
+
+type E6 struct{ F01 *inner }
+
+func TestStructPointer(t *testing.T) {
```
Comment (elithrar) [inline encoder_test.go]: Hooray for tests ;)

File: encoder.go [hunk 2]
```diff
@@ -62,6 +62,11 @@ func (e *Encoder) encode(v reflect.Value, dst map[string][]string) error {
 			continue
 		}
 
+		if v.Field(i).Type().Kind() == reflect.Ptr && v.Field(i).Elem().IsValid() && v.Field(i).Elem().Type().Kind() == reflect.Struct {
```
Comment (elithrar) [inline encoder.go]: Can you add a comment here that examples the reflect method chaining? The reflect API is a little obtuse, and thus making it clear why this calls what it does will help future maintainers & contributors.

(I understand it, but not everyone does!)

Comment (frozzare) [reply encoder.go]: Done 👍

**Comment by kisielk** [comment]
Can you also add a test case for when the inner struct pointer is nil?

**Comment by frozzare** [comment]
Sure. But it will return `null` because of this line https://github.com/gorilla/schema/blob/master/encoder.go#L124 that already exists of pointers.

File: encoder.go [hunk 3]
```diff
@@ -62,6 +62,12 @@ func (e *Encoder) encode(v reflect.Value, dst map[string][]string) error {
 			continue
 		}
 
+		// Encode struct pointer types if the field is a valid pointer and a struct.
+		if v.Field(i).Type().Kind() == reflect.Ptr && v.Field(i).Elem().IsValid() && v.Field(i).Elem().Type().Kind() == reflect.Struct {
```
Comment (kisielk) [inline encoder.go]: Please move the check to a function and change this to eg: `isValidStructPointer(v.Field(i))`

**Comment by frozzare** [comment]
@kisielk done

**Comment by kisielk** [comment]
Just had another thought, how does it behave with null pointers and `omitempty`? pointers to empty stucts and `omitempty`? Should probably have tests for both those cases as well.

**Comment by frozzare** [comment]
It already test nil pointers, added a test for `omitempty`

**Comment by kisielk** [comment]
That's what I meant by nil pointers and `omitempty`, as opposed to nil pointers or `omitempty` ;)

 Looks good though.
